docs: add event-bus, switchover and integrations (#1816)
openebs/mayastor: doc/design/control-plane.md

<br>

### Internal Communication

<br>

```mermaid
graph LR;
  subgraph Agents[" "]
    HACluster["HA Cluster Agent"]
    Core["Core Agent"]
    REST
  end

  subgraph StorageNode["Storage Node"]
    subgraph DataPlane["I/O Engine"]
      Nexus
      Replicas
      Pools
    end
  end

  subgraph AppNode["Application Node"]
    HANode["HA Node Agent"]
    CSINode["CSI Node Plugin"]
  end

  REST["Rest Server"] -->|gRPC| Core
  Core <-->|gRPC| DataPlane
  HACluster <-->|gRPC| HANode
  HACluster -->|gRPC| Core
  HANode -.->|UNIX Socket| CSINode
  CSIController -->|HTTP/REST| REST
```

<br>

> As you can see, there's many p2p connections between the different services.

Reviewer: ```suggestion
As we can see, there's many p2p connections between the different services.
```

## NATS & Fault management

> We used to use [NATS] as a message bus within mayastor as a whole, but later switched to [gRPC] for p2p communications. \

Reviewer 2: We can remove this IMO.

As of today, we use [NATS] for [events](./events.md) as async notifications via the [Publish/Subscribe Model](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern). Our current mechanism does not do any form of "consensus," retries, and the likes. Information transported over NATS will typically be error telemetry that is used to diagnose problems.

> > NOTE: We don't have any subscribers as things stand.

Reviewer 2: The cal-home stats aggregator is a subscriber IIRC.

Author: Indeed, done

## NATS & Fault management

> We used to use [NATS] as a message bus within mayastor as a whole, but as since switched for [gRPC] for p2p communications. \

Reviewer: ```suggestion
We used to use [NATS] as a message bus within Mayastor but later switched to [gRPC] for p2p communications. \
```

> As of today, we use [NATS] for [events](./events.md) via async notifications. Async in the sense that we send a message, but we do NOT wait for a reply. This mechanism does not do any form of "consensus," retries, and the likes. Information transported over NATS will typically be error telemetry that is used to diagnose problems. No work has started yet on this subject.

Reviewer: Maybe can we reword or put somewhere we follow publisher subscriber model but don't have any active subscriber at the moment, instead of explaining in what sense it's async?

Author: Reworded a bit

}
```

Read more about eventing [here](./events.md).
An up to date API of the event format can be fetched
 [here](https://github.com/openebs/mayastor-dependencies/blob/HEAD/apis/events/protobuf/v1/event.proto).

## Tracing and Telemetry

Tracing means different things at different levels. In this case, we are referring to tracing component boundary tracing.

> Tracing is by default implemented using open telemetry and, by default, we have provided a subscriber for [Jaeger]. From [Jaeger], the information can be

Reviewer 2: Nit: 'by default' is repeated

forwarded to, Elastic Search, Cassandra, Kafka, etc. In order to achieve full tracing support, all the [gRPC] requests and replies should add
`HTTP` headers such that we can easily tie them together in whatever tooling is used. This is standard practice but requires a significant amount of work.
The key reason is to ensure that all requests and responses pass along the headers, from `REST` to the scheduling pipeline.

We also need to support several types of transport and serialization mechanisms. For example, `HTTP/1.1 REST` requests to `HTTP/2 gRPC` request to
 a KV store operation to etcd. For this, we will use [Tower]. \

> [Tower] provides a not-so-easy to use an abstraction of request to response mapping.

Reviewer 2: Nit: Sentence seems incomplete, 'not-so-easy to use...'

Maybe, you mean, not-so-easy framework or interface or tooling?

Author: Done

```rust
let layer = TraceLayer::new_for_http().make_span_with(|request: &Request<Body>| {
  tracing::debug_span!(
    "HTTP",
    http.method = %request.method(),
    http.url = %request.uri(),
    http.status_code = tracing::field::Empty,
    // otel is a mandatory key/value
    otel.name = %format!("HTTP {}", request.method()),
    otel.kind = %SpanKind::Client,
    otel.status_code = tracing::field::Empty,
  )
})
```

> On the server-side we extra the trace id from the `HTTP` headers, and we inject it on the next call stack, which means it also gets eventually

Reviewer: ```suggestion
On the server-side we extract the trace id from the `HTTP` headers, and we inject it on the next call stack, which means it also gets eventually
```

> injected in the next transport hoop. Specifically for `REST`, this means we inject it on the inter-service `gRPC`. Again, we use the same `Service`

Reviewer: ```suggestion
injected in the next transport hop. Specifically for `REST`, this means we inject it on the inter-service `gRPC`. Again, we use the same `Service`
```
